src/OrleansRuntime/MembershipService/MembershipTableFactory.cs

```diff
@@ -0,0 +1,115 @@
+        }
+
+        internal async Task<IMembershipTable> GetMembershipTable()
+        {
+            if (membershipTable != null) return membershipTable;
+            using (this.initializationLock.LockAsync())
+            {
+                if (membershipTable != null) return membershipTable;
+
+                var globalConfig = this.serviceProvider.GetRequiredService<GlobalConfiguration>();
+                var livenessType = globalConfig.LivenessType;
+                IMembershipTable result;
+                if (livenessType.Equals(GlobalConfiguration.LivenessProviderType.MembershipTableGrain))
+                {
+                    var siloDetails = this.serviceProvider.GetRequiredService<SiloInitializationParameters>();
+                    if (siloDetails.Type == Silo.SiloType.Primary)
+                    {
+                        logger.Info(ErrorCode.MembershipFactory1, "Creating membership table grain");
+                        var catalog = this.serviceProvider.GetRequiredService<Catalog>();
+                        await catalog.CreateSystemGrain(
+                            Constants.SystemMembershipTableId,
+                            typeof(GrainBasedMembershipTable).FullName);
+                    }
+
+                    var grainFactory = this.serviceProvider.GetRequiredService<IInternalGrainFactory>();
+                    result =
+                        grainFactory.Cast<IMembershipTableGrain>(GrainReference.FromGrainId(Constants.SystemMembershipTableId));
+                }
+                else if (livenessType.Equals(GlobalConfiguration.LivenessProviderType.SqlServer))
+                {
+                    result = AssemblyLoader.LoadAndCreateInstance<IMembershipTable>(Constants.ORLEANS_SQL_UTILS_DLL, this.logger);
+                }
+                else if (livenessType.Equals(GlobalConfiguration.LivenessProviderType.AzureTable))
+                {
+                    result = AssemblyLoader.LoadAndCreateInstance<IMembershipTable>(Constants.ORLEANS_AZURE_UTILS_DLL, this.logger);
+                }
+                else if (livenessType.Equals(GlobalConfiguration.LivenessProviderType.ZooKeeper))
+                {
+                    result = AssemblyLoader.LoadAndCreateInstance<IMembershipTable>(
+                        Constants.ORLEANS_ZOOKEEPER_UTILS_DLL,
+                        this.logger);
+                }
+                else if (livenessType.Equals(GlobalConfiguration.LivenessProviderType.Custom))
+                {
+                    result = AssemblyLoader.LoadAndCreateInstance<IMembershipTable>(
+                        globalConfig.MembershipTableAssembly,
+                        this.logger);
+                }
+                else
+                {
+                    throw new NotImplementedException("No membership table provider found for LivenessType=" + livenessType);
+                }
+
+                await WaitForTableToInit(result);
```

Comment:
The table is now initialized before it's returned, whereas that was previously the responsibility of Silo. Now it is called from the `MembershipOracle.Start()` method.